espnet/nets/pytorch_backend/rnn/attentions.py

```diff
@@ -1376,6 +1413,18 @@ def forward(self, enc_hs_pad, enc_hs_len, dec_z, att_prev, out_prev, scaling=1.0
         if self.mask is None:
             self.mask = to_device(self, make_pad_mask(enc_hs_len))
         e.masked_fill_(self.mask, -float('inf'))
+
+        # NOTE(kan-bayashi): apply attention constraint introduced in DeepVoice3
+        if last_attended_idx is not None:
+            if batch != 1:
+                raise NotImplementedError("Batch attention constraining is not yet supported.")
+            backward_idx = last_attended_idx - backward_window
+            forward_idx = last_attended_idx + forward_window
+            if backward_idx > 0:
+                e[:, :backward_idx] = -float('inf')
+            if forward_idx < self.h_length:
+                e[:, forward_idx:] = -float('inf')
+
```

Comment:
Those three `if last_attended_idx` blocks looks very duplicated. How about a new function like this?
```python
def _apply_attention_constraint(e, last_attended_idx, forward_window, backward_window):
    if last_attended_idx is None:
        return
    if e.size(0) != 1:
        raise NotImplementedError("Batch attention constraining is not yet supported.")

    backward_idx = last_attended_idx - backward_window
    forward_idx = last_attended_idx + forward_window
    if backward_idx > 0:
        e[:, :backward_idx] = -float('inf')
    if forward_idx < e.size(1):
        e[:, forward_idx:] = -float('inf')
```

Reply:
Nice idea. I will change.

Reply:
I'm not sure inplace function is working correctly so I changed to `return e`.

Reply:
Done.